cpp/src/hash/spark_murmur_hash.cu

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2019-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/hashing.hpp>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/tabulate.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief Computes the hash value of a row in the given table.
+ *
+ * @tparam hash_function Hash functor to use for hashing elements.
+ * @tparam Nullate A cudf::nullate type describing whether to check for nulls.
+ */
+template <template <typename> class hash_function, typename Nullate>
+class device_spark_row_hasher {
```

Comment:
How should this be named?
- `device_spark_row_hasher`
- `spark_device_row_hasher`
- `device_row_spark_hasher`

Reply:
Do you need `device` in the name? How about just `spark_row_hasher` ?

Reply:
It's mirroring the functionality of `device_row_hasher` from here, so I think the parallel use of "device" is helpful: https://github.com/rapidsai/cudf/blob/a652ca918afb5ca38e9757d36df0cebc22ea296c/cpp/include/cudf/table/experimental/row_operators.cuh#L1393

Reply:
I'm leaning towards `spark_device_row_hasher` and will apply that for now.

Reply:
In this particular instance I agree with keeping `device` in the name.

Reply:
Having reviewed the code, @bdice I'm inclined to make this name _even more_ specific. Can we call it `spark_murmur_device_row_hasher` to clearly delineate that other hash functions wouldn't be supported?